myia/clone.py

```diff
@@ -0,0 +1,227 @@
+"""Graph cloning facility."""
+
+
+from typing import Any, Iterable, Dict, Union, Set, cast
+from myia.cconv import NestingAnalyzer
+from myia.info import About
+from myia.anf_ir import ANFNode, Apply, Parameter, Constant, Graph
+from myia.graph_utils import dfs
+from myia.anf_ir_utils import \
+    succ_incoming, exclude_from_set, \
+    is_parameter, is_constant
+from myia.utils import smap
+
+
+#################
+# Graph cloning #
+#################
+
+
+class GraphCloner:
```

Comment:
I'm having trouble wrapping my head around how this all fits together.

Also, I question the usefulness of having this as a class rather than a function.

Reply:
I think it's easier that way. Cloning a graph entails cloning every graph that's nested in it, and if you want to clone a set of graphs, it is best to do it all at once, because you need to change all calls to these graphs to calls to the cloned versions. A function version would have a lot of parameters and nested helpers, which isn't really any better in my opinion.